internal/cluster/discovery.go

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	pubcluster "github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/cluster/discovery"
+	"github.com/hazelcast/hazelcast-go-client/hzerrors"
+	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
+)
+
+const (
+	discoveryStrategyAdapterStarted = 1
+)
+
+type DiscoveryStrategyAdapter struct {
+	strategy    discovery.Strategy
+	lg          ilogger.LogAdaptor
+	translator  map[pubcluster.Address]pubcluster.Address
+	mu          *sync.RWMutex
+	opts        discovery.StrategyOptions
+	started     int32
+	usePublicIP bool
+}
+
+func NewDiscoveryStrategyAdapter(cfg pubcluster.DiscoveryConfig, lg ilogger.LogAdaptor) *DiscoveryStrategyAdapter {
+	if cfg.Strategy == nil {
+		panic("discovery strategy is required")
+	}
+	return &DiscoveryStrategyAdapter{
+		strategy:    cfg.Strategy,
+		translator:  map[pubcluster.Address]pubcluster.Address{},
+		usePublicIP: cfg.UsePublicIP,
+		mu:          &sync.RWMutex{},
+		lg:          lg,
+		opts: discovery.StrategyOptions{
+			Logger:      lg,
+			UsePublicIP: cfg.UsePublicIP,
+		},
+	}
+}
+
+func (d *DiscoveryStrategyAdapter) Addresses(ctx context.Context) ([]pubcluster.Address, error) {
```

Comment:
As far as I see, loadAddresses method in java returns private addresses regardless of the usePublicIp config. See https://github.com/hazelcast/hazelcast/blob/7c3894b1474dba56648cf10ad8441615486fcdea/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/discovery/RemoteAddressProvider.java#L42

Reply:
Are you sure, there's this line: https://github.com/hazelcast/hazelcast/blob/7c3894b1474dba56648cf10ad8441615486fcdea/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/discovery/RemoteAddressProvider.java#L44
```
privateToPublic = getAddresses.call();
```

Reply:
That `.call()` calls this method:
 https://github.com/hazelcast/hazelcast/blob/7c3894b1474dba56648cf10ad8441615486fcdea/hazelcast/src/main/java/com/hazelcast/client/impl/clientside/ClusterDiscoveryServiceBuilder.java#L161

but it does not check the usePublicIp config as well.